### PR TITLE
Support HCR with java agent

### PR DESCRIPTION
**Pull request by skinny121**
This adds the ability to hot swap mixin code.

To run add -Dmixin.hotSwap=true and -javaagent:\<path to the mixin jar\> to vm options.

I got the path to the mixin jar by adding the following to the Sponge/SpongeCommon/gradle/minecrft.gradle file
 ```groovy
task debugArguments << {
    configurations.compile.each {
        String absolutePath = it.absolutePath
        if(absolutePath.contains("mixin")) {
            println("-javaagent:"+absolutePath)
        }
    }
}
``` 
## Java Agent and HCR

A java agent is a mechanism to hook into the class loading phase of the java runtime. With a java agent you can register a transformer and it will be called on any time a class is defined and have the ability to change the bytecode of the loading class. HCR(Hot code replacement) works by the debugger calling a method in the com.sun.jdi api, this then redefines a class and calls the transformers.

## Implementation

When the mixin's are initialised they tell the java agent about themselves so that it can load a fake class with the same name as the mixin into a custom classloader. This is done because the mixin class is needed to be loaded as the debugging interface won't try to reload if the class if it isn't loaded. 

When a class is being transformed by the mixin transformer it also saves the original bytecode of the class so that the mixins can be reapplied on top of the 'clean' bytecodes.

So when a class is redefined it first checks if there it is a mixin, if so it gets all of the target classes of that mixin and for each reapplies the mixins for that class on top of the 'clean' bytecode. Then returns the same bytecode of the fake mixin class, because then this allows for the addition of Shadow fields and method as well as the add/remove Overwrite methods.

## Limitations

Because java doesn't allow adding/removing methods/fields in HCR, certain changes won't work like adding/removing normal methods/field and adding/removing injections(but should be able to change the injections target). 

Currently changes the target classes and priority of the mixin isn't supported but they should be achievable.

### CODE REVIEW

**Comment by bloodmc** [comment]
@skinny121 This is great! I just tested it and confirm it works perfectly in eclipse debugger.

@Mumfrey please get this merged when possible as this will save us a TON of time debugging.

**Comment by Mumfrey** [comment]
Nice, I like it. Crude but effective! This should work for the immediate use cases though I have some minor worries about it, see below.

@bloodmc actually asked about this the other day and I told him it would be a pain to implement, but given that HCR is restricted to method bodies only that should mitigate basically all of the issues which would arise from full-on hot swapping, so he will be very happy!

The main issue that remains is the downstream transformer chain. Mixin does its best to ensure that it's last in the transformer chain (via the proxy transformers) but that's not a contractual guarantee and in practice could lead to weird behaviour, since any transformers which run after the mixin transformer will have their changes erased when a mixin is hot-swapped. This probably isn't a problem for Sponge, because we don't have any other transformers, but this project isn't soley consumed by Sponge.

In the short term that's only a worry and not a show-stopper. The main thing that needs to change before this is pulled is the handling of `reloadMixin`. Currently this calls `validate` on the `MixinInfo`, which actually nowadays ought to be called maybe `prepareAndValidate` because it does more than just run the validation. This can be easily fixed by just adding a semaphore to `MixinInfo` to ensure that the certain tasks which should only run once only run once. I also think the behaviour of `reloadMixin` in handling validation failures is wrong:

It looks like you've just copied the logic from `postInitialise`, but failures at this point are still *before* any actual mixin application, so dropping the mixin from the config makes sense since it's invalid. However I think a better behaviour for `reloadMixin` would be to throw an error which signals the hotswap has failed so that the IDE can report this back (current behaviour on validation failure would be that it just logged a message to the console and after that it would just not hot swap any more).

Other than that though, I don't see any immediate problems with this other than some style/naming things and I'll comment those inline.

File: src/agent/java/org/spongepowered/tools/agent/MixinAgent.java [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+public class MixinAgent implements IHotSwap {
+
+    /**
+     * Class file transformer that re-transforms mixin's target classes if the
+     * mixin has been redefined.
+     */
+    private class Transformer implements ClassFileTransformer {
+
+        @Override
+        public byte[] transform(ClassLoader loader,
```
Comment (Mumfrey) [inline src/agent/java/org/spongepowered/tools/agent/MixinAgent.java]: Only break the line where necessary, at 150 chars.

File: src/agent/java/org/spongepowered/tools/agent/MixinAgent.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+     * Initialize the java agent
+     *
+     * This will be called automatically if the jar is in a -javaagent java
+     * command line argument
+     *
+     * @param arg Ignored
+     * @param instrumentation Instance to use to transform the mixins
+     */
+    public static void premain(String arg, Instrumentation instrumentation) {
+        init(instrumentation);
```
Comment (Mumfrey) [inline src/agent/java/org/spongepowered/tools/agent/MixinAgent.java]: Missing `this`.

File: src/agent/java/org/spongepowered/tools/agent/MixinAgent.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+     * Initialize the java agent
+     *
+     * This will be called automatically if the java agent is loaded after jvm
+     * startup
+     *
+     * @param arg Ignored
+     * @param instrumentation Instance to use to re-define the mixins
+     */
+    public static void agentmain(String arg, Instrumentation instrumentation) {
+        init(instrumentation);
```
Comment (Mumfrey) [inline src/agent/java/org/spongepowered/tools/agent/MixinAgent.java]: Also missing `this`.

File: src/agent/java/org/spongepowered/tools/agent/MixinClassLoader.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+    /**
+     * Add a fake mixin class
+     *
+     * @param name Name of the fake class
+     */
+    void addMixinClass(String name) {
+        logger.debug("Mixin class " + name + " added to class loader");
+        try {
+            byte[] bytes = materialize(name);
```
Comment (Mumfrey) [inline src/agent/java/org/spongepowered/tools/agent/MixinClassLoader.java]: Missing `this`.

File: src/agent/java/org/spongepowered/tools/agent/MixinClassLoader.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+     */
+    private Map<String, byte[]> targets = new HashMap<String, byte[]>();
+
+    /**
+     * Add a fake mixin class
+     *
+     * @param name Name of the fake class
+     */
+    void addMixinClass(String name) {
+        logger.debug("Mixin class " + name + " added to class loader");
```
Comment (Mumfrey) [inline src/agent/java/org/spongepowered/tools/agent/MixinClassLoader.java]: Unqualified access to `logger`.

File: src/main/java/org/spongepowered/asm/mixin/transformer/MixinConfig.java [hunk 6]
```diff
@@ -504,6 +508,32 @@ public boolean hasMixinsFor(String targetClass) {
...
+     */
+    public List<String> reloadMixin(String mixinClass, byte[] bytes) {
+        for (Iterator<MixinInfo> iter = this.mixins.iterator(); iter.hasNext();) {
+            MixinInfo mixin = iter.next();
+            if (mixin.getClassName().equals(mixinClass)) {
+                mixin.reloadMixin(bytes);
+                try {
+                    mixin.validate();
+                } catch (Exception e) {
+                    this.logger.error(e.getMessage(), e);
```
Comment (Mumfrey) [inline src/main/java/org/spongepowered/asm/mixin/transformer/MixinConfig.java]: This was the catch I was talking about, potentially have this throw an error which can be caught and signalled to the instrumentation API.

File: src/main/java/org/spongepowered/asm/mixin/transformer/MixinTransformer.java [hunk 7]
```diff
@@ -421,6 +453,24 @@ public void audit() {
...
+     * @param mixinClass Name of the mixin
+     * @param bytes New bytecode
+     * @return List of classes that need to be updated
+     */
+    public List<String> reload(String mixinClass, byte[] bytes) {
+        if (this.lock.getDepth() > 0) {
+            throw new MixinApplyError("Cannot reload mixin if re-entrant lock entered");
+        }
+        List<String> targets = new ArrayList<String>();
+        for (MixinConfig config:this.configs) {
```
Comment (Mumfrey) [inline src/main/java/org/spongepowered/asm/mixin/transformer/MixinTransformer.java]: spacing around `:`.

File: src/agent/java/org/spongepowered/tools/agent/MixinClassLoader.java [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+        return this.targets.get(name);
+    }
+
+    /**
+     * Generates the simplest possible class that is instantiable
+     *
+     * @param name Name of the generated class
+     * @return Bytecode of the generated class
+     */
+    private static byte[] materialize(String name) {
```
Comment (Mumfrey) [inline src/agent/java/org/spongepowered/tools/agent/MixinClassLoader.java]: Use english spellings old chap.

File: src/agent/java/org/spongepowered/tools/agent/MixinClassLoader.java [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+import org.spongepowered.asm.lib.Type;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Class loader that is used to load fake mixin classes so that they can be
+ * re-defined.
+ */
+class MixinClassLoader extends ClassLoader {
```
Comment (Mumfrey) [inline src/agent/java/org/spongepowered/tools/agent/MixinClassLoader.java]: I would prefer if this had a more obvious name, `MixinClassLoader` is pretty general sounding but this isn't really related to the core behaviour of the library so maybe something more specific like `MixinAgentClassLoader`.

**Comment by skinny121** [comment]
I have changed MixinInfo validation so that if any error that occurs during validation, that it just reverts all the state back to before the validation starts. This means if someone tries to HCR a invalid mixin they can just try again with a correct mxin and it will work, no state corruption. This also makes it easy to check for changes that aren't allowed while reloading and report back a sensible message. As far as can tell I have eliminated all global visible modifications that occur during validation of the reloaded mixin.

For errors I have changed it so that the agent returns an illegal class file, this means that It will inform the user that something did go wrong and then they can look in the log, but the message the IDE displays isn't sensible. This is a limitation of the instrumentation API, that it has no mechanism to handle custom error messages, but for its primary use cases it doesn't make sense as you don't expect instrumentation tool to flat out reject a class. This problem could be solved if and when there is some sort of IDE support.

**Comment by Mumfrey** [comment]
That makes sense, it sucks that there's no nice way to "talk back" to the IDE via instrumentation but this approach is way better than just logging an error and having the mixin fail altogether. I have a few tiny formatting things for you to clean up then this can be merged. Will comment those inline.

File: src/agent/java/org/spongepowered/tools/agent/MixinAgent.java [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+            }
+            byte[] mixinBytecode = MixinAgent.classLoader.getFakeMixinBytecode(classBeingRedefined);
+            if (mixinBytecode != null) {
+                MixinAgent.logger.info("Redefining mixin " + className);
+                List<String> targets;
+                try {
+                    targets = MixinAgent.this.classTransformer.reload(className.replace('/', '.'), classfileBuffer);
+                } catch (MixinReloadException e) {
+                    MixinAgent.logger.error("mixin " + e.getMixinInfo() + " cannot be reloaded, needs a restart to be applied: " + e.getMessage());
+                    return ERROR_BYTECODE;
```
Comment (Mumfrey) [inline src/agent/java/org/spongepowered/tools/agent/MixinAgent.java]: Unqualified access to static field.

File: src/agent/java/org/spongepowered/tools/agent/MixinAgent.java [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+                } catch (Throwable th) {
+                    // catch everything as otherwise it is ignored
+                    MixinAgent.logger.error("Error while finding targets for mixin " + className, th);
+                    return ERROR_BYTECODE;
```
Comment (Mumfrey) [inline src/agent/java/org/spongepowered/tools/agent/MixinAgent.java]: here too.

File: src/agent/java/org/spongepowered/tools/agent/MixinAgent.java [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+                }
+                for (String target : targets) {
+                    String targetName = target.replace('/', '.');
+                    MixinAgent.logger.debug("Re-transforming target class " + target);
+                    try {
+                        Class<?> targetClass = Launch.classLoader.findClass(targetName);
+                        byte[] targetBytecode = MixinAgent.classLoader.getOriginalTargetBytecode(targetName);
+                        if (targetBytecode == null) {
+                            MixinAgent.logger.error("Target class " + targetName + " hasn't registered its bytecode");
+                            return ERROR_BYTECODE;
```
Comment (Mumfrey) [inline src/agent/java/org/spongepowered/tools/agent/MixinAgent.java]: gonna stop here because there's more than I thought, but you get the gist.

**Comment by Mumfrey** [comment]
@skinny121 are you happy with this or are there any further tweaks you want to make before it's merged?

**Comment by skinny121** [comment]
I have nothing I want to change, so you can merge this.

**Comment by Mumfrey** [comment]
Merged this and cleaned up the few tiny remaining checkstyle things. Thanks for your work on this it's much appreciated.

**Comment by ryantheleach** [comment]
```
Error opening zip file or JAR manifest missing : 
Error occurred during initialization of VM
agent library failed to init: instrument
```

Is there something missing in the instructions for this?

**Pull request merged**